server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java

```diff
@@ -0,0 +1,490 @@
+                }
+                entries[i] = new Entry(elem, matchingResponse, matchingResponse.id);
+            }
+        }
+
+        // sanity check
+        if (nextInternalId != -1) {
+            throw new IllegalArgumentException("id mismatch");
+        }
+
+        // now link entries together in history order, and add a sentinel node at the beginning
+        Entry first = new Entry(null, null, -1);
+        Entry lastEntry = first;
+        for (Entry entry : entries) {
+            lastEntry.next = entry;
+            entry.prev = lastEntry;
+            lastEntry = entry;
+        }
+
+        return first;
+    }
+
+    enum EventType {
+        INVOCATION,
+        RESPONSE
+    }
+
+    public static class Event implements Serializable {
+        public final EventType type;
+        public final Object value;
+        public final int id;
+
+        public Event(EventType type, Object value, int id) {
+            this.type = type;
+            this.value = value;
+            this.id = id;
+        }
+
+        @Override
+        public String toString() {
+            return "Event{" +
+                "type=" + type +
+                ", value=" + value +
+                ", id=" + id +
+                '}';
+        }
+    }
+
+    static class Entry {
+        final Event event;
+        final Entry match; // null if current entry is a response, non-null if it's an invocation
+        final int id; // internal id, distinct from Event.id
+        Entry prev;
+        Entry next;
+
+        Entry(Event event, Entry match, int id) {
+            this.event = event;
+            this.match = match;
+            this.id = id;
+        }
+
+        // removes this entry from the surrounding structures
+        void lift() {
+            prev.next = next;
+            next.prev = prev;
+            match.prev.next = match.next;
+            if (match.next != null) {
+                match.next.prev = match.prev;
+            }
+        }
+
+        // reinserts this entry into the surrounding structures
+        void unlift() {
+            match.prev.next = match;
+            if (match.next != null) {
+                match.next.prev = match;
+            }
+            prev.next = this;
+            next.prev = this;
+        }
+    }
+
+
+    private static class Cache {
+        private final Map<Object, Set<FixedBitSet>> largeMap = new HashMap<>();
+        private final LongObjectHashMap<Set<Object>> smallMap = new LongObjectHashMap<>();
+        private final Map<Object, Object> internalizeStateMap = new HashMap<>();
+        private final Map<Set<Object>, Set<Object>> statePermutations = new HashMap<>();
+        public boolean add(Object state, FixedBitSet bitSet) {
+            return addInternal(internalizeStateMap.computeIfAbsent(state, k -> state), bitSet);
+        }
+
+        private boolean addInternal(Object state, FixedBitSet bitSet) {
```

Comment:
We should add our own BitSet instead, since both BitSet and FixedBitSet has unnecessary memory overheads.